### PR TITLE
Add provoking vertex mode

### PR DESCRIPTION
**Pull request by AMD-dwang**
Provoking vertex mode specifies which vertex of a primitive is the _provoking
vertex_, this impacts which vertex's "flat" VS outputs are passed to the PS.
We will need this to fix the barycentric coordinate.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Review by s-perron — COMMENTED** [review]
From what I can tell this does not have any impact on the elf that is generated.  Is there an issue that describes how this will be used and what work is upcoming?

At this point I cannot tell if this value should be included in the pipeline hash. Does it need to be in the hash for relocatable shaders as well?

For what it is, a new input for which we do nothing, I think this is fine.

**Comment by amdvlk-admin** [comment]
### Test summary for commit ccfcb154e1cd4dd1130fda4904fc8a59012e3b49
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev bcec463efe5260776d486a5e3da0c549bc0a75d2</li>
<li>SPVGEN: dev 580ade71d630a304805e9468ef22ad263c37d137</li>
<li>XGL: dev 0429e80c4fa09141362e586d785743ee80cfc551</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

File: lgc/builder/InOutBuilder.cpp [hunk 1]
```diff
@@ -105,6 +105,7 @@ Value *InOutBuilder::CreateReadPerVertexInput(Type *resultTy, unsigned location,
   Value *even = nullptr;
 
   auto primType = getPipelineState()->getPrimitiveType();
+  auto provokingMode = getPipelineState()->getRasterizerState().provokingVertexMode;
```
Comment (linqun) [inline lgc/builder/InOutBuilder.cpp:108]: please use same variable name in whole change. (provokingMode -> provokingVertexMode)

File: include/vkgcDefs.h [hunk 2]
```diff
@@ -793,6 +794,10 @@ struct GraphicsPipelineBuildInfo {
                                   ///  matches the sample pattern used by the rasterizer when rendering
                                   ///  with this pipeline.
     uint8_t usrClipPlaneMask;     ///< Mask to indicate the enabled user defined clip planes
+
+    VkProvokingVertexModeEXT provokingVertexMode; ///< Specifies which vertex of a primitive is the _provoking
```
Comment (linqun) [inline include/vkgcDefs.h:798]: please move usrClipPlaneMask before numSamples  to group all 32bit member together.

**Review by linqun — COMMENTED** [review]
looks good to me for others.

File: include/vkgcDefs.h (hunk 2 above)
Comment (AMD-dwang) [reply include/vkgcDefs.h:798]: Fixed, thanks.

File: lgc/builder/InOutBuilder.cpp (hunk 1 above)
Comment (AMD-dwang) [reply lgc/builder/InOutBuilder.cpp]: Fixed, thanks.

**Review by amdrexu — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0a7e923ceee3dee5ef12ad7feff5f089e306e356
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 39f8940199e60c44d4211cf8165dfd12876316fa</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 0a0a4ae4ab062d31fcd68863c35952967a8060ee</li>
<li>SPVGEN: dev 6c2f36bcf0f8ff7ba8d20c8b789ca6346d1e16b1</li>
<li>XGL: dev c41a825d1ea17c108e8fb123f082357f849e7823</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev c20ba7614ecdc78b3d99601423783100835200c1</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228192)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 33955/57048 (59.5%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23093/57048 (40.5%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 33797/57048 (59.2%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 23250/57048 (40.8%)</li><li>Warnings: 1/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 34085/57048 (59.7%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 22963/57048 (40.3%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 35882/57048 (62.9%)</li>
<li>Failed: 0/57048 (0.0%)</li>
<li>Not Supported: 21166/57048 (37.1%)</li><li>Warnings: 0/57048 (0.0%)</li>
</ul></details>


</ul></details>

**Review by amdrexu — APPROVED** [review]

**Pull request merged**